File: problem-solving/_lc-81-medium-search-in-rotated-sorted-array-ii/1-binary-search.py
"""
https://leetcode.com/problems/search-in-rotated-sorted-array-ii/

Complexity
----------
Time:
  Best case: O(logn)
  Worst case: O(n) -> When all elements are same and we are searching for some other element.
Space: O(1)

Explanation: https://leetcode.com/problems/search-in-rotated-sorted-array-ii/solutions/1890199/c-algorithm-binary-search
"""

class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        left = 0
        right = len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return True

            # Case: [3,1,2,3,3,3,3] -> can't decide which way to go.
            if nums[left] == nums[mid] == nums[right]:
                left += 1
                right -= 1
            elif nums[left] <= nums[mid]:
                if nums[left] <= target <= nums[mid]:
                    right = mid - 1
                else:
                    left = mid + 1
            else:
                if nums[mid] <= target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
        return False

File: problem-solving/lc-154-hard-find-minimum-in-rotated-sorted-array-ii/1-binary-search.go
/*
https://leetcode.com/problems/find-minimum-in-rotated-sorted-array-ii/

Complexity
----------
Time:
  Best Case: O(logn)
  Worst Case: O(n)
Space: O(1)
*/

func findMin(nums []int) int {
    left, right := 0, len(nums) - 1
    for left < right {
        mid := (left + right) / 2
        if nums[mid] > nums[right] {
            left = mid + 1
        } else if nums[mid] < nums[right] {
            right = mid
        } else {
            right -= 1
        }
    }
    return nums[left]
}
